design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,238 @@
+
+The `ASSERT DATA` statement asserts that data exists at an optionally provided offset (or otherwise 
+sequential based on the last `ASSERT DATA`) with the given values. If the `ASSERT DATA` does not
+specify certain columns they will be ignored (allowing users to assert only subset of the columns
+match what they expect).
+
+The values will be created in the same way that `INSERT INTO` creates values:
+```sql
+ASSERT DATA source_name [OFFSET offset] ( { column_name } [, ...] ) VALUES ( value [, ...] );
+```
+
+The `ASSERT NO DATA` statement will allow users to ensure that no more data exists in a specified
+source:
+```sql
+ASSERT NO DATA source_name;
+```
+
+The `ASSERT (TABLE | STREAM)` statement asserts that the given stream or table has the specified
+columns and physical properties.
+```sql
+ASSERT (STREAM | TABLE) source_name ( { column_name data_type [[PRIMARY] KEY] } [, ...] )
+    WITH (property_name = expression [, ...] );
+```
+
+#### Considered Alternative
+
+It is possible to use a meta directive (see below) in order to drive asserts as well. This was
+considered and rejected because it would require additional parsing and more testing-specific code
+to convert to expressions and pass into the codegen, which diverges from the motivating principle 
+of minimizing testing-only code.
+
+### Meta Directives
+
+Some functionality cannot be addressed in standard SQL syntax and will instead be supported using
+meta directives. Some examples of these directives:
+
+- The `--@expected.error` and `--@expected.message` directives will indicate that the below test 
+    expects an exception to be thrown with the specified message and type.
+- The `--@topic.denylist` directive will ensure that any topics matching the denylist will not
+    be present at the end of the test execution.
+    
+### Language Gaps
+
+There are some features that we need to implement in ksqlDB's SQL language in order to put YATT
+on par with the existing testing tools. Some of those are outlined here:
+
+- _Windowed Keys_: when we have structured key support, we can `ASSERT DATA` and just provide the
+    struct as the value. Until then, windowed keys will either be unsupported or stringified.
+- _Tombstones_: we will implement `DELETE FROM table WHERE key = value` syntax to allow tombstones
+    to be inserted into tables, both for ksqlDB and the testing tool.
+- _Null Values for Streams_: insert `null` values into streams will not be supported for the first
+    iteration of YATT. Eventually, if we want to support this, we can add a directive like
+    `--@null.value: topic key` to produce a `null` valued record into a specific topic/source.
+- _Other Unsupported Inserts_: there are other types of inserts that ksqlDB doesn't allow for at the
+    moment (e.g. `enum` support or binary formats). This is only somewhat a regression from the 
+    existing QTT tests, so we may or may not decide to support it in YATT.
+    
+### Error Reporting & UX
```

Comment:
One thing I'd love to get into QTT, but the large existing set of tests prohibits, is for tests to fail if anything unexpected ends up in the processing logger.  It would be great if we can add this, from the start, to YATT.

Only errors are written to the processing logger.  Test cases should not, generally, be generating such errors. Silently ignoring them means we/users can write tests that pass, yet would spam the processing log (and likely app log) if used in production.  We already know queries that generate a lot of processing errors have terrible performance.

YATT could initially just fail the test case if there are entries in the processing logger. (We'll have to work out how to demux errors when running tests in parallel - probably need to add some kind of UUID or similar somehow).  Later we can add `ASSERT LOG` or similar to assert entries in the processing log, allowing for negative testing. Such negative testing is an important part of QTT.

Reply:
This seems like a good idea, might make migrating tests a little harder so we'd need to be able to disable this unless we wanted to block on adding the correct asserts or fixing up the tests.

Also we could use `ASSERT DATA` on the processing log stream, which should be registered by ksql anyway.

Reply:
Yeah, something on-by-default.  Can follow the same pattern of other tests in that they should fail if there is unexpected output byond what the test expected.  Unexpected output on the processing log = failed test.